text/0000-macro-at-most-once-rep.md

```diff
@@ -0,0 +1,166 @@
+- Feature Name: macro-at-most-once-rep
+- Start Date: 2018-01-17
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+
+Summary
+-------
+
+Add a repetition specifiers to macros to repeat a pattern at most once: `$(pat)?`.  `?` behaves like `+` or `*` but represents at most one repetition of `pat`.
+
+Motivation
+----------
+
+There are two specific use cases in mind:
+
+1. Any macro rule with optional parts. Currently, you just have to write two rules and possibly have one "desugar" to the other.
+
+   ```rust
+   macro_rules! foo {
+     (do $b:block) => {
+       $b
+     }
+     (do $b1:block and $b2:block) => {
+       foo!($b1)
+       $b2
+     }
+   }
+   ```
+
+   Under this RFC, one would simply write:
+
+   ```rust
+   macro_rules! foo {
+     (do $b1:block $(and $b2:block)?) => {
+       $b1
+       $($b2)?
+     }
+   }
+   ```
+
+2. Trailing commas. It's kind of infuriating that the best way to make a rule tolerate trailing commas TMK is to create another identical rule that has a comma at the end:
```

Comment:
Stylistic suggestions:
+ `It's kind of infuriating` feels a bit strong and subjective
+ Expand `TMK` => to my knowledge. (even better: remove it all together).